### ISSUE TITLE
Figure 4.2

### ISSUE THREAD

**Issue by KrTis** [astroML/astroML#82]
Hi,
upon running the code for Book Figure 4.2 on Ubuntu,  Python returned an error:  'GMM' object has no attribute 'eval' for `logprob, responsibilities =M_best.eval(x)`.
To solve the problem, I replaced `M_best.eval(x)` (line 85) with: 
`M_best.score_samples(x.reshape((-1,1)))`
and `M_best.predict_proba(x)` (line 110) with: 
`p = M_best.predict_proba(x.reshape((-1,1)))`

I'm using scikit-learn 0.17

Best regards,
KT

**Comment by jakevdp** [comment]
Hi,
Yes, there have been some syntax changes in scikit-learn over the last year or so. I'm hoping to update all these pieces soon – thank you for the bug report!

**Comment by joel-roediger** [comment]
Hi Jake,

I encountered similar problems when trying to recreate Fig. 4.2 today.  I am attaching my patchwork solution, which features several changes from the version used for the book.  I would particularly like to draw your attention to lines 29-32, where I recreated the sample dataset from scratch.  The latest version of scikit-learn returns an error on line 34, saying that a fit has not been done yet.  Also, GaussianMixture seems to prefer receiving datasets in the form of vertical stacks, not just a horizontal array.  Furthermore, there does not seem to be an equivalent method in the present version of GaussianMixture to output responsibilities (line 75); I have not found a work-around for this yet.  Finally, I found that the AIC and BIC preferred a two-component model rather than the expected three.  Testing showed that I could reproduce the expected result by shifting the location of the first Gaussian to -2.  I am still puzzled why the new version of GaussianMixture struggles in this sense.
[fig4.2.py.txt](https://github.com/astroML/astroML/files/997677/fig4.2.py.txt)
